[MRG] Update contributing guidelines and issue templates (#655)
jupyter/repo2docker: .github/ISSUE_TEMPLATE/feature_request.md

assignees: ''

---
<!-- Thank you for contributing. These HTML commments will not render in the issue, but you can delete them once you've read them if you prefer! -->

### Proposed change
<!-- Use this section to describe the feature you'd like to be added. -->

### Alternative options
<!-- Use this section to describe alternative options and why you've decided on the proposed feature above. -->

### Who would use this feature?
<!-- Describe the audience for this feature. This information will affect who chooses to work on the feature with you. -->

> ### How much effort will adding it take?

Reviewer: One idea here (feel free to ignore !), would be to ask folks to say _where_ they think that effort will be targeted, rather than just how much. It could help to avoid answers like "3 hours" and would make sense if contributors are more or less comfortable with different parts of the code base.

Reviewer 2: I like this. Though also see the earlier discussion where we concluded that it almost doesn't matter what you write here. The goal is to get people thinking about the fact that there is no free lunch and not to get a realistic estimate (which is super hard even for practiced teams)

Author: (see below for a suggested addition)

> <!-- Try to estimate how much work adding this feature will require. This information will affect who chooses to work on the feature with you. -->

Author: ```suggestion
<!-- Try to estimate how much work adding this feature will require. Be as specific as you can (while recognising that its really hard to predict the future!). This information will affect who chooses to work on the feature with you. -->
```

Author: @betatim & @emdupre - Do you think this addition is useful? I like the point your making....but I wonder if the prompt is too long?

I'm very much on the fence with this, so just emoji vote this :+1: or :-1: (or suggest a different one!) and I'll go with what you think 👾

Reviewer 2: "If in doubt shorter" wins for me on this one.

Author: Ok. Not going to merge this now. Let’s see how the prompt goes....

assignees: ''

---
<!-- Thank you for contributing. These HTML commments will not render in the issue, but you can delete them once you've read them if you prefer! -->

### Proposed change
<!-- Use this section to describe the feature you'd like to be added. -->

### Alternative options
<!-- Use this section to describe alternative options and why you've decided on the proposed feature above. -->

### Who would use this feature?
<!-- Describe the audience for this feature. This information will affect how work is assigned across the maintenance and contributing team. -->

> ### How much time will adding it take?

Reviewer 2: I wouldn't ask this question. Even in my day job where we are a well oiled team of people who have worked together for a while on a small code base we are terrible at estimating efforts required. Estimates seem to come in two categories: spot on (for work that takes "no time") and way off (for work where you discover that actually it is much harder than you thought).

Author: Can it be rephrased to effort? The goal is to get the person opening the issue to think about how much labour they're asking for. (Not to actually allocate time for the reasons you say!)

Happy to drop though!

Author: (Made a suggestion below)

Reviewer 2: I like the idea of getting people to think about how much work it is. As an example https://github.com/jupyter/repo2docker/pull/637 where I think everyone thought it would be a quick job but turned out to be a bit more work and required more experience :-/

Conclusion: I like the idea, I wonder how we deal with the fact that it is super hard/imprecise. Though maybe that doesn't matter as the whole goal is to get people to think about the topic. In which case we don't actually care about the answer (just don't tell anyone we don't care :-) )

Author: Yes - exactly! I don't think the _actual_ answer matters in this case, just that folks have thought about it 😆 (and say something at least within the realms of possibility).

> ### How much time will adding it take?

Author: ```suggestion
### How much effort will adding it take?
```

Reviewer 2: 👍 , I can't press commit on the suggestion but if I could I would.

> <!-- Try to estimate how much work adding this feature will require. This information will affect how work is assigned across the maintenance and contributing team. -->

Author: ```suggestion
<!-- Try to estimate how much work adding this feature will require. This information will affect who in the the maintenance and contributing team chooses to work on the feature with you. -->
```

Author: Should be same as change above, lets keep the conversation there ☝️ to make it easier to follow.

### Who can do this work?

> <!-- What skills are needed? Who can be recruited to add this feature? This information will affect how work is assigned across the maintenance and contributing team. -->

Author: ```suggestion
<!-- What skills are needed? Who can be recruited to add this feature? This information will affect who in the the maintenance and contributing team chooses to work on the feature with you. -->
```

Reviewer 2: I'd take this or go another step further with:

> What skills are needed? Who can be recruited to add this feature? This information will affect who chooses to work on the feature with you.

Author: Gotcha - I think I'm trying to have them realise that maintaining software is a task that volunteers do. I guess I'm worried that people don't even _think_ about these tasks, so by explicitly naming them its bringing those considerations to the fore.

I totally agree that "maintainers" and "contributors" (hopefully) overlap! Definitely don't want to imply that they're two totally separate groups of people.

I'm going to leave this thread to see if a few other folks comment, but I'm fine with changing it to "who chooses to work on the feature with you" if there are a few votes in that direction 👾

Reviewer 3: +1 on "who chooses" vs. "assigned"

Reviewer: I do like "who chooses" here :+1:

> ### Thank you maintainers!

Author: Remove? 😭 

Give me an emoji vote on getting rid of this section :+1: to keep, :-1: to remove.

Reviewer 3: I'm also leaning towards removing this, since this issue template is already a bit long and so it may not have the effect we want. I'd love to find another way to let people give positive feedback to the community - maybe an on-going thread in Discourse or something like this?

Reviewer 2: Nods. I think the thanks has to be genuine and heartfelt. Otherwise it can be a negative thing as well. (Remember that guy who read that pop-sci article on how people like hearing their own name, that it is a good way to remember people's names and so ends up using your name every second sentence?)

This makes it hard to channel into anything "standard" because ... well if it is standard then it doesn't mean anything, it just becomes transactional (like saying thanks when I pass you the butter at dinner). So maybe a reminder or pointer to nudge people towards the right behaviour?
